File: php/test1.php
<?php

$INPUT_CSV_FILE = "./data/sales.csv";
$OUTPUT_CSV_FILE = "./php/salesWithTaxes.csv";
$BENCHMARK_CSV_FILE = "./output/benchmarks.csv";
$GLOBALS['outputPath'] = $BENCHMARK_CSV_FILE;
$TAX_PERCENT = 0.05;

$GLOBALS['logBench'] = false;

function benchmark(string $tag, float $timeStart, float $memoryStart)
{
  $timeUsage = number_format((microtime(true) - $timeStart) * 1000, 2) . " ms";
  $memoryUsage = number_format((memory_get_usage() - $memoryStart) / 1024 / 1024, 2) . " mb";
  if ($GLOBALS['logBench']) {
    echo ("-------- BENCHMARK " . strtoupper($tag) . " RESULTS --------" . "\n");
    echo ("time: " . $timeUsage . " memory: " . $memoryUsage . "\n");
    echo ("-------- END BENCHMARK  RESULTS --------" . "\n" . "\n");
  }

  $fp = fopen($GLOBALS["outputPath"], 'a');
  fputcsv($fp, ["php", $tag, $timeUsage, $memoryUsage]);
  fclose($fp);
}

$timeStart = microtime(true);
$memoryStart = memory_get_usage();

$rows = [];
$handle = fopen($INPUT_CSV_FILE, "r"); // open in readonly mode
while (($row = fgetcsv($handle)) !== false) {
  $rows[] = $row;
}
fclose($handle);

benchmark("read file", $timeStart, $memoryStart);

$timeStart = microtime(true);
$memoryStart = memory_get_usage();

$headers = array_slice($rows, 0, 1);
$sales = array_slice($rows, 1);

$salesWithTaxes = [];

foreach ($sales as $row) {
  // parse string money format text into float for calculation
  $price = floatval(substr($row[2], 1));
  $addTax = number_format($price + $price * $TAX_PERCENT, 2);
  $salesWithTaxes[] = [$row[0], $row[1], '$' . $addTax];
}

$fullData = array_merge($headers, $salesWithTaxes);

benchmark("process data",  $timeStart, $memoryStart);

$timeStart = microtime(true);
$memoryStart = memory_get_usage();

$fp = fopen($OUTPUT_CSV_FILE, 'w');
foreach ($fullData as $row) {
  fputcsv($fp, $row);
}
fclose($fp);

benchmark("write file",  $timeStart, $memoryStart);

File: go/test1.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const INPUT_CSV_FILE = "./data/sales.csv"
const OUTPUT_CSV_FILE = "./go/salesWithTaxes.csv"
const BENCHMARK_CSV_FILE = "./output/benchmarks.csv"
const TAX_PERCENT = 0.05

var logBench bool = false

type Bench struct {
	language  string
	benchType string
	timeUse   time.Duration
	memoryUse string
}

func getMemoryUsage() uint64 {
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)
	return mem.Alloc / 1024 / 1024 // Convert from bytes to mb
}

func benchmark(tag string) func() {
	timeStart := time.Now()
	memoryStart := getMemoryUsage()
	return func() {
		timeEnd := time.Now()
		timeUsage := timeEnd.Sub(timeStart)
		memoryUsage := fmt.Sprintf("%.2vmb", getMemoryUsage()-memoryStart)
		if logBench {
			fmt.Println("-------- BENCHMARK ", strings.ToUpper(tag), " RESULTS --------")
			fmt.Println("time: ", timeUsage, " memory: ", memoryUsage)
			fmt.Println("-------- END BENCHMARK  RESULTS --------")
			fmt.Println()
		}
		benchmarkFile, err := os.OpenFile(BENCHMARK_CSV_FILE, os.O_APPEND, 0644)
		if err != nil {
			fmt.Println(err)
		}
		writer := csv.NewWriter(benchmarkFile)
		writer.Comma = '\t'
		defer writer.Flush()
		bench := []string{fmt.Sprintf("golang,%s,%s,%s", tag, timeUsage, memoryUsage)}
		writer.Write(bench)
	}
}

func main() {
	readBench := benchmark("read file")
	inputFile, err := os.Open(INPUT_CSV_FILE)
	if err != nil {
		fmt.Println(err)
	}
	reader := csv.NewReader(inputFile)
	records, _ := reader.ReadAll()
	readBench()

	processBench := benchmark("process data")
	sales := records[1:]

	for _, sale := range sales {
		price, err := strconv.ParseFloat(sale[2][1:], 32)
		if err != nil {
			fmt.Println(err)
		}
		addTax := fmt.Sprintf("$%.2f", price+price*TAX_PERCENT)

		sale[2] = addTax
	}
	processBench()
	writeBench := benchmark("write file")
	outputFile, err := os.Create(OUTPUT_CSV_FILE)
	if err != nil {
		fmt.Println(err)
	}
	writer := csv.NewWriter(outputFile)
	defer writer.Flush()
	writer.WriteAll(records)
	writeBench()
}
